File: mqttclients/dotnet/MQTTnet.Client.Extensions/TelemetryConsumer.cs
namespace MQTTnet.Client.Extensions;

public class TelemetryConsumer<T>
{
    private readonly IMqttClient _mqttClient;
    private readonly IMessageSerializer _serializer;
    private readonly string _topicPattern;
    public Action<TelemetryMessage<T>>? OnTelemetryReceived { get; set; }

    public TelemetryConsumer(IMqttClient mqttClient, IMessageSerializer serializer, string topicPattern)
    {
        _mqttClient = mqttClient;
        _serializer = serializer;
        _topicPattern = topicPattern;
    }

    public async Task StartAsync()
    {
        _mqttClient.ApplicationMessageReceivedAsync += async m =>
        {
            await Task.Yield();
            string topic = m.ApplicationMessage.Topic;

            MqttTopicFilterCompareResult res = MqttTopicFilterComparer.Compare(topic, _topicPattern);
            if (res == MqttTopicFilterCompareResult.IsMatch)
            {
                string[] segments = topic.Split('/');
                TelemetryMessage<T> msg = new()
                {
                    ClientIdFromTopic = segments[1],
                    Payload = _serializer.FromBytes<T>(m.ApplicationMessage.PayloadSegment.Array!)
                };
                OnTelemetryReceived?.Invoke(msg);
            }
        };
        await _mqttClient.SubscribeAsync(_topicPattern, Protocol.MqttQualityOfServiceLevel.AtLeastOnce);
    }
}

File: mqttclients/ts/mqttjsClientExtensions/src/telemetryConsumer.ts
import {
    IPublishPacket,
    MqttClient
} from 'mqtt';
import { matches as mqttMatches } from 'mqtt-pattern';
import {
    IMessageSerializer,
    TelemetryMessage
} from './';

export class TelemetryConsumer<T> {
    private mqttClient: MqttClient;
    private serializer: IMessageSerializer;
    private topicPattern: string;
    public onTelemetryReceived: (msg: TelemetryMessage<T>) => Promise<void>;

    public constructor(mqttClient: MqttClient, serializer: IMessageSerializer, topicPattern: string) {
        this.mqttClient = mqttClient;
        this.serializer = serializer;
        this.topicPattern = topicPattern;
    }

    public async startAsync(): Promise<void> {
        this.mqttClient.on('message', (topic: string, payload: Buffer, _packet: IPublishPacket) => {
            if (mqttMatches(this.topicPattern, topic)) {
                const segments = topic.split('/');
                const msg = new TelemetryMessage<T>(segments[1], this.serializer.fromBytes<T>(payload));

                void this.onTelemetryReceived(msg);
            }
        });

        await this.mqttClient.subscribeAsync(this.topicPattern, { qos: 1 });
    }
}